### PR TITLE
CP-35758: write sysctl configuration to file

### PR DESCRIPTION
**Pull request by alexhimmel**
Signed-off-by: alexhimmel <alex.ac@qq.com>

### CODE REVIEW

File: autocertkit/utils.py [hunk 1]
```diff
@@ -1844,12 +1844,19 @@ def init_vifs_ip_addressing(session, vm_ref, vifs_info):
                       (vm_ref, device))
 
 
+def modify_interface_rp_filter(mip, iface):
```
Comment (lindig) [inline autocertkit/utils.py:1847]: I think the name could be more specific. "Modify" is quite general. What is `rp_filter=0` doing? This also goes for the comment in the code. What exactly does this? That this is writing to a file is obvious but not what is accomplishes.

I don't see the need for `-e`; an `echo` without `-e` and `\n` would do the same thing but if this how it's done consistently in this code base it's fine.

Is writing to `/etc/sysctl.d/99-customer.conf` enough for it to take effect? Or does this require some form of reload? Has this been tested?

**Review by lindig — APPROVED** [review]

**Comment by DeliZhangX** [comment]
It's able to fix the VLAN 0 test issue indeed.
While I am thinking another better solution, because `reset_arp` is still the proper place to apply the settings. In ACK, currently IperfTest calls `reset_arp`  but other ping test case does not yet. We may consider to move all `reset_arp` call just after VM(vifs) creating.
Besides each ethx, we need to write the global setting, e.g. `net.ipv4.conf.all.arp_ignore=1`
let's discuss details tomorrow.

**Comment by DeliZhangX** [comment]
You can do like below

- Add function  `set_kernel_parameter(key, value, mip=None)` in plugins\autocertkit which is to set key=value for VM or Dom0 /etc/sysctl.conf also load it by `SYSCTL -p`
- Optimize `reset_arp` to use the function 
- In `reset_arp` execute command `ip -o link | awk '{if($2 ~ "eth.+:") print $2}' | sed 's/:$//'` to get all ethx of VM and call `set_kernel_parameter`
- In test `VLANTestClass.test_vlan_high_port`  `call_ack_plugin(session, 'reset_arp', ...)` for the 2 VMs before ping test running.

And optional, as we can improve in future
- Define `PingTest` class like `IperfTest`, move all `reset_arp`  calls to the class or VM vifs created, then sysctl.conf settings will become the default rule for all test cases.

**Comment by alexhimmel** [comment]
Copy that, will do

File: autocertkit/network_tests.py [hunk 2]
```diff
@@ -426,6 +426,7 @@ def test_vlan_high_port(self, session):
         # Make certain the VMs are available
         for vm_ref in [vm1_ref, vm2_ref]:
             check_vm_ping_response(session, vm_ref, get_context_vm_mip(vm_ref))
+            call_ack_plugin(session, 'reset_arp', {'vm_ref': vm_ref, 'mip': get_context_vm_mip(vm_ref)})
```
Comment (DeliZhangX) [inline autocertkit/network_tests.py:429]: Should put this call before line 426

File: autocertkit/utils.py [hunk 3]
```diff
@@ -1842,8 +1842,8 @@ def init_vifs_ip_addressing(session, vm_ref, vifs_info):
         else:
             log.debug("Init VM %s device %s as default dhcp" %
                       (vm_ref, device))
-
-
+ 
+
```
Comment (DeliZhangX) [inline autocertkit/utils.py:1846]: redundant space chars

File: plugins/autocertkit [hunk 4]
```diff
@@ -1203,6 +1203,18 @@ def remove_route(session, args):
     return json_dumps("OK")
 
 
+@log_exceptions
+def set_kernel_parameter(key, value, mip=None):
+    """
+    Delete all line which contains key in config file, and then write key=value to file,
+    Will take effect after reload(sysctl --sytstem) or reboot
+    """
+    conf_file = "/etc/sysctl.d/99-customer.conf"
```
Comment (DeliZhangX) [inline plugins/autocertkit:1212]: could be `/etc/sysctl.d/ack.conf`, or just modify `/etc/sysctl.conf` directly as it the last config file to load.

File: plugins/autocertkit [hunk 5]
```diff
@@ -1203,6 +1203,18 @@ def remove_route(session, args):
...
+    cmd = "sed -i '/%s/d' %s; echo -e '%s=%s\n'>>%s" % (key, conf_file, key, value, conf_file)
```
Comment (DeliZhangX) [inline plugins/autocertkit:1213]: `'/%s/d'` to `'/^%s=/d'`
`echo -e '%s=%s\n'` to `echo '%s=%s'`

File: plugins/autocertkit [hunk 6]
```diff
@@ -1265,16 +1277,20 @@ def reset_arp(session, args):
         # In previous DLVM (CentOS 5), rp_filter is 0. If don't set in new DLVM, then Multicast test
         # will be failed using interfaces in same one subnet
         # Change global setting
-        cmd = "%s net.ipv4.conf.all.rp_filter=0; %s net.ipv4.conf.default.rp_filter=0" \
-              % (SYSCTL, SYSCTL)
-        call = [SSH, vm_m_ip, cmd]
-        make_local_call(call)
+        set_kernel_parameter("net.ipv4.conf.all.rp_filter", 0, vm_m_ip)
+        set_kernel_parameter("net.ipv4.conf.default.rp_filter", 0, vm_m_ip)
```
Comment (DeliZhangX) [inline plugins/autocertkit:1281]: could modify line 1265 as well by call `set_kernel_parameter`

**Comment by alexhimmel** [comment]
@DeliZhangX 
Thank you for your help to review.
I have modified code as your comment above, and has tested with xenrt.

**Review by DeliZhangX — APPROVED** [review]

**Pull request merged**